validator.go

```diff
@@ -0,0 +1,261 @@
+package jwt
+
+import (
+	"crypto/subtle"
+	"time"
+)
+
+// validator is the core of the new Validation API. It is automatically used by
+// a [Parser] during parsing and can be modified with various parser options.
+//
+// Note: This struct is intentionally not exported (yet) as we want to
+// internally finalize its API. In the future, we might make it publicly available.
+type validator struct {
+	// leeway is an optional leeway that can be provided to account for clock skew.
+	leeway time.Duration
+
+	// timeFunc is used to supply the current time that is needed for
+	// validation. If unspecified, this defaults to time.Now.
+	timeFunc func() time.Time
+
+	// verifyIat specifies whether the iat (Issued At) claim will be verified.
+	// According to https://www.rfc-editor.org/rfc/rfc7519#section-4.1.6 this
+	// only specifies the age of the token, but no validation check is
+	// necessary. However, if wanted, it can be checked if the iat is
+	// unrealistic, i.e., in the future.
+	verifyIat bool
+
+	// expectedAud contains the audience this token expects. Supplying an empty
+	// string will disable aud checking.
+	expectedAud string
+
+	// expectedIss contains the issuer this token expects. Supplying an empty
+	// string will disable iss checking.
+	expectedIss string
+
+	// expectedSub contains the subject this token expects. Supplying an empty
+	// string will disable sub checking.
+	expectedSub string
+}
+
+// CustomClaims represents a custom claims interface, which can be built upon the integrated
+// claim types, such as map claims or registered claims.
+type CustomClaims interface {
```

Comment:
I'm struggling a bit with this one. Can we ship `v5` without this interface, and consider this in minor release?

The caller already has access to the claims, and they can do whatever validation they need. Why do we need an extra interface only to call the `CustomValidation()` method on behalf of the caller.

Reply:
This was sort of a compromise since my feeling was that currently a lot of people are overriding the `Valid` method on their custom claim to add additional piece of validation in the "correct" place. Since claims do not have the valid method anymore, I wanted users to give some "canonical" place to do additional validation. I believe this is a valid use case and much cleaner than first doing the "regular" claims validation and then if that succeeds doing an additional one, with extra error handling, etc.

Having this here in the example is quite nice imho.
https://github.com/golang-jwt/jwt/blob/3a9ee81ba34953206aea4c4ebbd5d78fa2afa339/example_test.go#L119-L130

The question is, whether this interface could actually be unexported, so we are more flexible with naming it, etc.?

Reply:
My worry is this is a potential footgun for users because the indirection of second interface.

If users are embedding custom claims, one could argue it should be their responsibuility to validate those claims. But I do see the argument for having a canonical place to do validation.

My general thinking around this package (I could be wrong) is to do the bare minimum to verify the token and validate the claims with an API that eliminates as many footguns as possible.

Reply:
Another option could be to inline the type?

```go
// Finally, we want to give the claim itself some possibility to do some
// additional custom validation based on a custom function
type customValidator interface {
	Validate() error
}
cvt, ok := claims.(customValidator)
if ok {
	if err := cvt.Validate(); err != nil {
		vErr.Inner = err
		vErr.Errors |= ValidationErrorClaimsInvalid
	}
}
```

Reply:
Maybe I am missing the point about your "indirection". Are you concerned about the naming? The way you did it in the example also looks good to me.

Reply:
See https://github.com/golang-jwt/jwt/pull/234/commits/1e16f55059b4ef3ce614e7d5b22bc347cb253ef3

Reply:
I was going back and forth on how to best expose this to users. After thinking about it some more, it feels like we should expose this interface and document its use-case, the way you had it makes it discoverable and allows a compile-time assertion of a custom claims.

```go
type CustomValidator interface {
	Validate() error
}
```

Wdyt?

Reply:
Either one is fine for me. Lets keep it inlined for now (for the release candidate) and document its use case in the upcoming docs.

Reply:
Ugh, on the other hand this feels a bit confusing, because we have a `(v *validator) Validate(...)` method that takes a `Claims` interface .. and now that interface needs to satisfy a `CustomValidator` interface with a `Validate() error` method.

Users need to create a concrete type, that embeds a type or implements one that satisfies a `Claims` interface. And then it also needs to have a `Validate() error` to satisfy a `CustomValidator` interface.

Is this complicated, and can it be made easier. That's kind of what I'm trying to explore.

Reply:
I tried this out, and this interface helps glue together the claims in the package and any custom validation, maybe this is the least-worst option?

```go
type MyCustomClaims struct {
	Foo string `json:"foo"`
	jwt.RegisteredClaims
}

var _ jwt.ClaimsValidator = (*MyCustomClaims)(nil)

func (m MyCustomClaims) Validate() error {
	if m.Foo != "bar" {
		return errors.New("must be foobar")
	}
	return nil
}
```

Reply:
> I tried this out, and this interface helps glue together the claims in the package and any custom validation, maybe this is the least-worst option?
> 
> ```go
> type MyCustomClaims struct {
> 	Foo string `json:"foo"`
> 	jwt.RegisteredClaims
> }
> 
> var _ jwt.ClaimsValidator = (*MyCustomClaims)(nil)
> 
> func (m MyCustomClaims) Validate() error {
> 	if m.Foo != "bar" {
> 		return errors.New("must be foobar")
> 	}
> 	return nil
> }
> ```

Looks good. I implemented this in #274